### PR TITLE
Sips 10 & 14 improve delegation powers

### PR DESCRIPTION
**Pull request by hav-noms**
https://sips.synthetix.io/sips/sip-10
https://sips.synthetix.io/sips/sip-14

### CODE REVIEW

File: contracts/interfaces/IDelegateApprovals.sol [hunk 1]
```diff
@@ -0,0 +1,14 @@
+pragma solidity 0.4.25;
+
+
+interface IDelegateApprovals {
+    function rateForCurrency(bytes32 currencyKey) external view returns (uint);
```
Comment (jacko125) [inline contracts/interfaces/IDelegateApprovals.sol]: maybe dont need this one here

Comment (hav-noms) [reply contracts/interfaces/IDelegateApprovals.sol]: oh whoops... how'd that get in there

File: contracts/Synthetix.sol [hunk 2]
```diff
@@ -269,22 +269,6 @@ contract Synthetix is ExternStateToken, MixinResolver {
         return _transferFrom_byProxy(messageSender, from, to, value);
     }
 
-    function issueSynths(uint amount) external optionalProxy {
```
Comment (justinjmoses) [inline contracts/Synthetix.sol]: Right yeah it makes sense to start moving these out now.

Comment (hav-noms) [reply contracts/Synthetix.sol]: @justinjmoses @jacko125  I'm not 100% certain this is a good way to go just yet. I am concerned about composability. 
Before integration could just rely on the synthetix proxy to call issue /burn on now the requirement would be to integrate the address resolver so they can always be pointed to the current issuer.  Maybe I'm overthinking it but if the docs explain this well enough could avoid people hard coding the. Issuer address only to have it deprecated when we upgrade it. Also, nobody has integrated with Issue and burn yet but delegation should open that up.

@justinjmoses does this also mean we remove synthetix.exchange() for Exchanger.exchange()?

Comment (justinjmoses) [reply contracts/Synthetix.sol]: Discussed offline. Summary: leave it all in `Synthetix` for now and put together a SIP to gather all these user functions into one kind of forwarder contract that will also emit these events.

File: contracts/Exchanger.sol [hunk 3]
```diff
@@ -26,6 +27,7 @@ contract Exchanger is MixinResolver {
     bytes32 private constant CONTRACT_EXRATES = "ExchangeRates";
     bytes32 private constant CONTRACT_SYNTHETIX = "Synthetix";
     bytes32 private constant CONTRACT_FEEPOOL = "FeePool";
+    bytes32 private constant CONTRACT_DELEGATEAPPROVALS = "DelegateApprovals";
 
     bytes32[24] private addressesToCache = [CONTRACT_EXCHANGESTATE, CONTRACT_EXRATES, CONTRACT_SYNTHETIX, CONTRACT_FEEPOOL];
```
Comment (justinjmoses) [inline contracts/Exchanger.sol]: it needs to be added in here as well

File: contracts/Issuer.sol [hunk 4]
```diff
@@ -28,6 +29,7 @@ contract Issuer is MixinResolver {
     bytes32 private constant CONTRACT_EXCHANGER = "Exchanger";
     bytes32 private constant CONTRACT_SYNTHETIXSTATE = "SynthetixState";
     bytes32 private constant CONTRACT_FEEPOOL = "FeePool";
+    bytes32 private constant CONTRACT_DELEGATEAPPROVALS = "DelegateApprovals";
     bytes32 private constant CONTRACT_ISSUANCEETERNALSTORAGE = "IssuanceEternalStorage";
 
     bytes32[24] private addressesToCache = [
```
Comment (justinjmoses) [inline contracts/Issuer.sol]: ibid

File: contracts/Synthetix.sol [hunk 5]
```diff
@@ -273,18 +273,34 @@ contract Synthetix is ExternStateToken, MixinResolver {
...
 
+    function issueMaxSynthsOnBehalf(address issueForAddress) external optionalProxy {
+        return issuer().issueMaxSynthsOnBehalf(issueForAddress, messageSender);
+    }
+
     function burnSynths(uint amount) external optionalProxy {
         return issuer().burnSynths(messageSender, amount);
     }
 
+    function burnSynthsOnBehalf(address burnForAddress, uint amount) external {
```
Comment (justinjmoses) [inline contracts/Synthetix.sol]: not optionalProxy?

File: contracts/Issuer.sol [hunk 6]
```diff
@@ -150,9 +167,18 @@ contract Issuer is MixinResolver {
         _internalBurnSynths(from, debtToRemoveAfterSettlement, existingDebt, totalSystemValue);
     }
 
+    function burnSynthsToTargetOnBehalf(address burnForAddress) external {
```
Comment (justinjmoses) [inline contracts/Issuer.sol]: onlySynthetix and missing an arg?

File: contracts/Exchanger.sol (hunk 3 above)
Comment (hav-noms) [reply contracts/Exchanger.sol]: for exchange sip-14

File: contracts/Synthetix.sol (hunk 5 above)
Comment (hav-noms) [reply contracts/Synthetix.sol]: Yes good catch thanks

**Comment by justinjmoses** [comment]
@Clinton regarding the test flakes - I ran this 100 separate times locally and couldn't reproduce it. That being said, calculating the interest rate down to the second and expecting ganache's timestamps to match your expected probably isn't the best approach - can you instead request the time before and after the fastForward and use that amount of time to calculate the expected interest?

**Comment by jacko125** [comment]
@justinjmoses  Makes sense to use the time before and after from ganache to calculate expected interest instead of relying up to the second. I can update those tests.

File: test/contracts/DelegateApprovals.js [hunk 7]
```diff
@@ -1,9 +1,11 @@
 const DelegateApprovals = artifacts.require('DelegateApprovals');
+const { onlyGivenAddressCanInvoke } = require('../utils/setupUtils');
+const { toBytes32 } = require('../../.');
```
Comment (justinjmoses) [inline test/contracts/DelegateApprovals.js]: @hav-noms can you add the only mutative function expectation checker to this test suite?

File: test/contracts/Issuer.js [hunk 8]
```diff
@@ -1531,4 +1538,107 @@ contract('Issuer (via Synthetix)', async accounts => {
...
+				await assert.revert(synthetix.burnSynthsToTargetOnBehalf(authoriser, { from: delegate }));
+			});
+		});
+
+		it('should approveIssueOnBehalf for account1', async () => {
+			const authoriser = account1;
+			const delegate = account2;
+
+			await delegateApprovals.approveIssueOnBehalf(delegate, { from: authoriser });
+			const result = await delegateApprovals.canIssueFor(authoriser, delegate);
```
Comment (justinjmoses) [inline test/contracts/Issuer.js]: this is the only test for canIssueFor - we need to test the failure case as well

File: test/contracts/Exchanger.js [hunk 9]
```diff
@@ -100,7 +100,13 @@ contract('Exchanger (via Synthetix)', async accounts => {
...
 			abi: exchanger.abi,
 			ignoreParents: ['MixinResolver'],
-			expected: ['settle', 'setExchangeEnabled', 'setWaitingPeriodSecs', 'exchange'],
+			expected: [
+				'settle',
+				'setExchangeEnabled',
+				'setWaitingPeriodSecs',
+				'exchange',
+				'exchangeOnBehalf',
+			],
```
Comment (justinjmoses) [inline test/contracts/Exchanger.js]: we need a test for both the success and failure cases of `exchangeOnBehalf`

File: test/contracts/Issuer.js (hunk 8 above)
Comment (justinjmoses) [reply test/contracts/Issuer.js]: right ok i see the reverts on the failure cases for issue and burn above

File: test/contracts/Exchanger.js (hunk 9 above)
Comment (jacko125) [reply test/contracts/Exchanger.js]: Correct, been added now.

**Pull request merged**